Inject OSImageURL from CVO into templated MachineConfigs

cgwalters opened the pull request:
This injects the `OSImageURL` into the "base"
config (e.g. `00-worker`, `00-master`).  This differs from
previous pull requests which made it a separate MC, but that
adds visual noise and will exacerbate renderer race conditions.

jlebon commented on pkg/controller/template/render.go:
```diff
@@ -233,7 +249,12 @@ func generateMachineConfigForName(config *RenderConfig, role, name, path string)
-	return MachineConfigFromIgnConfig(role, name, ignCfg), nil
+	mcfg := MachineConfigFromIgnConfig(role, name, ignCfg)
+	if osUpdatesEnabledForRole(config, role) {
+		mcfg.Spec.OSImageURL = config.OSImageURL
```
Let's maybe also log something in the `else` branch here so it's clear why it's not picking it up for some pools?

cgwalters replied:
See https://github.com/openshift/machine-config-operator/pull/348

cgwalters commented:
Moving [discussion of bootstrap](https://github.com/openshift/machine-config-operator/pull/324#issuecomment-459827557) over here.

Now that I think about it more...before or after this lands, in fact nothing will be reacting to the osimageurl.  So we should be able to land the installer PR now.

abhinavdahiya commented on pkg/apis/machineconfiguration.openshift.io/v1/types.go:
```diff
@@ -141,6 +141,10 @@ type ControllerConfigSpec struct {
 
+	// By default, operating system updates are enabled for a pool.  This
+	// flag allows disabling them.  If any entry is "*" it will match all pools.
+	OSUpdatesDisabledForPools []string `json:"osUpdatesDisabledForPools,omitempty"`
```
use label selectors.

cgwalters replied:
How do the `labels.Everything()` vs `labels.Nothing()` serialize?  We want to support both here right?

runcom commented on pkg/controller/template/render.go:
```diff
@@ -239,7 +254,15 @@ func generateMachineConfigForName(config *RenderConfig, role, name, path string)
+	osUpdatesEnabled, err := osUpdatesEnabledForRole(config, role)
+	if err != nil {
+		return nil, err
+	} else if osUpdatesEnabled {
```
No need for this else, you're returning on the branch aboce anyway

cgwalters replied:
I'm confused, the other one is usual error handling?  Are you saying it'd e.g. be clearer like this?

```
diff --git a/pkg/controller/template/render.go b/pkg/controller/template/render.go
index 40aa9a8..97cf0e4 100644
--- a/pkg/controller/template/render.go
+++ b/pkg/controller/template/render.go
@@ -258,7 +258,8 @@ func generateMachineConfigForName(config *RenderConfig, role, name, path string)
 	osUpdatesEnabled, err := osUpdatesEnabledForRole(config, role)
 	if err != nil {
 		return nil, err
-	} else if osUpdatesEnabled {
+	}
+	if osUpdatesEnabled {
 		mcfg.Spec.OSImageURL = config.OSImageURL
 	}
 
```

runcom replied:
Yup, that's how it's usually done when you have an if branch which returns

runcom commented:
> My idea here is that anyone who wants to test things out can
> `oc edit controllerconfig` and empty that out. Another useful
> thing would be to change it to e.g. `{"master"}` and test OS updates
> on workers without affecting the master.

to clear out confusion, I believe this comment no longer stands

cgwalters commented:
/test e2e-aws-op

cgwalters commented:
```
I0208 18:18:26.188194    5356 daemon.go:516] Bootstrap pivot required
I0208 18:18:26.188204    5356 update.go:655] Updating OS to registry.svc.ci.openshift.org/ci-op-l08p2kyf/stable@sha256:21eba43a81fd1a6f9e114b2b957e398aeeb75cfd9a8f74c3b62fb714ba42e23c
I0208 18:18:26.188212    5356 run.go:13] Running: /bin/pivot registry.svc.ci.openshift.org/ci-op-l08p2kyf/stable@sha256:21eba43a81fd1a6f9e114b2b957e398aeeb75cfd9a8f74c3b62fb714ba42e23c
pivot version 0.0.2 (f55cf7b5d1b832ad3fecfff1aca09aa0f6969fc7)
```
...
```
I0208 18:20:58.639732    4944 start.go:52] Version: 3.11.0-586-g95a3071d-dirty
I0208 18:20:58.640449    4944 start.go:88] starting node writer
I0208 18:20:58.648801    4944 run.go:22] Running captured: chroot /rootfs rpm-ostree status --json
I0208 18:20:58.690876    4944 daemon.go:155] Booted osImageURL: registry.svc.ci.openshift.org/ci-op-l08p2kyf/stable@sha256:21eba43a81fd1a6f9e114b2b957e398aeeb75cfd9a8f74c3b62fb714ba42e23c (47.291)
I0208 18:20:58.692434    4944 daemon.go:227] Managing node: ip-10-0-174-241.ec2.internal
I0208 18:21:12.781634    4944 start.go:146] Calling chroot("/rootfs")
I0208 18:21:12.781706    4944 daemon.go:404] In bootstrap mode
I0208 18:21:13.551441    4944 daemon.go:432] Current+desired config: worker-095a512e0970e036a0f262d657a327da
I0208 18:21:13.551544    4944 daemon.go:520] No bootstrap pivot required; unlinking bootstrap node annotations
I0208 18:21:13.554818    4944 daemon.go:547] Validated on-disk state
I0208 18:21:13.554889    4944 daemon.go:579] In desired config worker-095a512e0970e036a0f262d657a327da
I0208 18:21:13.554993    4944 start.go:165] Starting MachineConfigDaemon
I0208 18:21:13.555048    4944 daemon.go:248] Enabling Kubelet Healthz Monitor
```

```
                    "worker": "3 out of 3 nodes have updated to latest configuration worker-095a512e0970e036a0f262d657a327da"
```

:confetti_ball:

cgwalters commented:
Ah but here's the next problem, the `machine-os-content` needs to be updated:

We went from `47.308` to `47.291`:

```
Upgraded:
  nss-altfiles 0-2.atomic.git20131217gite2a80593.el7 -> 2.18.1-11.el7
  ostree 2019.1-1.el7_6 -> 2019.1.5-6649032a375238255052a43adb8bc56faac989ca.8cbd7fc123ad6d6e4e8216211aee6f7dd6264886.el7
  ostree-grub2 2019.1-1.el7_6 -> 2019.1.5-6649032a375238255052a43adb8bc56faac989ca.8cbd7fc123ad6d6e4e8216211aee6f7dd6264886.el7
  pivot 0.0.2-0.1.el7 -> 0.0.2.11-f1ed664ed83e73268464e81019f213366d961bb2
  rpm-ostree 2019.1-3.atomic.el7 -> 2019.1.4-fa5be441b177a40b285ed1abc539c6f7770ab231.091833c72cefe9fcbb3af2b42dd07d8a8c9f63d2.el7
  rpm-ostree-libs 2019.1-3.atomic.el7 -> 2019.1.4-fa5be441b177a40b285ed1abc539c6f7770ab231.091833c72cefe9fcbb3af2b42dd07d8a8c9f63d2.el7
Downgraded:
  atomic-openshift-clients 4.0.0-0.164.0.git.0.88cca3f.el7 -> 4.0.0-0.150.0.git.0.f39ab66.el7
  atomic-openshift-hyperkube 4.0.0-0.164.0.git.0.88cca3f.el7 -> 4.0.0-0.150.0.git.0.f39ab66.el7
  atomic-openshift-node 4.0.0-0.164.0.git.0.88cca3f.el7 -> 4.0.0-0.150.0.git.0.f39ab66.el7
  cri-o 1.12.5-5.rhaos4.0.git9076a33.el7 -> 1.12.5-2.rhaos4.0.gitd4191df.el7
  glusterfs 3.12.2-40.el7 -> 3.12.2-32.el7
  glusterfs-client-xlators 3.12.2-40.el7 -> 3.12.2-32.el7
  glusterfs-fuse 3.12.2-40.el7 -> 3.12.2-32.el7
  glusterfs-libs 3.12.2-40.el7 -> 3.12.2-32.el7
  redhat-release-coreos 4.0-20180515.0.atomic.el7.0 -> 0-4749e1e9959e9dcb53804ed103dfde64c813ecd7.el7
  runc 1.0.0-58.dev.rhaos4.0.git2abd837.el7 -> 1.0.0-57.dev.git2abd837.el7
Removed:
  ostree-fuse-2019.1-1.el7_6.x86_64
Added:
  bubblewrap-0.3.1.6-94147e233fe200d1fe43a9a18c52475188b22798.el7.centos.x86_64
  ostree-libs-2019.1.5-6649032a375238255052a43adb8bc56faac989ca.8cbd7fc123ad6d6e4e8216211aee6f7dd6264886.el7.x86_64
```

ashcrow commented:
```
--- FAIL: TestMCDeployed (306.89s)
	mcd_test.go:146: machine config didn't result in file being on any worker: timed out waiting for the condition
```

cgwalters commented:
Yeah, that's because we were also doing an OS update and pivot failed due to needing a new rpm-ostree.

ashcrow commented:
Updated `rpm-ostree` built by @jlebon should now be available.

ashcrow commented:
/test e2e-aws-op

cgwalters commented:
/hold cancel

This should be ready to go.

cgwalters commented:
Though let's rerun the tests a few times to see if we're introducing any new flakes.

/test e2e-aws-op

cgwalters commented:
> time="2019-02-12T16:54:36Z" level=debug msg="Still waiting for the cluster to initialize: Cluster operator machine-config is reporting a failure: Failed when progressing towards 3.11.0-607-ge92cf6ea-dirty because: error syncing: timed out waiting for the condition during syncRequiredMachineConfigPools: error pool master is not ready. status: (total: 3, updated: 0, unavailable: 1)"

Ah, looks like pivot failed:

> error: opendir(/var/lib/containers/storage/overlay/bbfa59102dda40133ac5a26eb6a5877f7b859f8aaff46223988e914e6b499fe3/merged/srv/repo): No such file or directory
F0212 16:46:56.911922    5558 run.go:62] rpm-ostree: exit status 1
E0212 16:46:56.912465    5520 daemon.go:384] Fatal error checking initial state of node: Checking initial state: Failed to run pivot: exit status 255

ashcrow commented:
@jlebon is the error on `/var/lib/containers/storage/overlay/bbfa59102dda40133ac5a26eb6a5877f7b859f8aaff46223988e914e6b499fe3/merged/srv/repo` related to https://github.com/openshift/pivot/pull/31?

cgwalters commented:
This now blocks on https://github.com/openshift/machine-config-operator/pull/335

cgwalters commented:
Sweet, that merged.  Rebased :surfing_man:  and let's see what the tests say.

/hold

So we can run the e2e tests a few times and gather results.

ashcrow commented:
https://github.com/openshift/machine-config-operator/pull/335 is now merged

cgwalters commented:
In [that e2e-aws run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws-op/429) we got:

```
level=fatal msg="failed to initialize the cluster: Cluster operator machine-config is reporting a failure: Failed when progressing towards 3.11.0-616-g5c68396a-dirty because: error syncing: timed out waiting for the condition during syncRequiredMachineConfigPools: error pool master is not ready. status: (total: 3, updated: 0, unavailable: 1)"
```

And the root cause is ye olde:

```
I0212 20:55:50.891321   16585 daemon.go:906] While getting MachineConfig master-36db45cb5e256d4ffb546f52c6ef3ff4, got: machineconfigs.machineconfiguration.openshift.io "master-36db45cb5e256d4ffb546f52c6ef3ff4" not found. Retrying...
```

But...without https://github.com/openshift/installer/pull/1189 it's going to be a pain to debug.

I may look at redoing that PR.  Hmm.  Or maybe at least we can extract what data we're passing in `bootkube.sh` in artifacts.

/test e2e-aws-op

cgwalters commented:
/test e2e-aws

cgwalters commented:
One thing I don't understand is that https://github.com/openshift/release/pull/2813 only seems to kick in sometimes.  In [that e2e-aws-op job](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws-op/441) we didn't get archived `machineconfigs.json`.  But in other PRs we did.

cgwalters commented:
I may have figured it out...https://github.com/openshift/machine-config-operator/pull/363/commits/d58765b4cddca60ce55de903d03e0bff13f9ffcb

cgwalters commented:
/test e2e-aws-op

ashcrow commented:
/test e2e-aws-op

cgwalters commented:
/test e2e-aws-op

ashcrow commented:
Prometheus and proxy errors

/test e2e-aws

cgwalters commented:
In [that e2e-aws run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1711) it looks like we successfully updated[1] all masters and workers!

A lot of "watch closed before UntilWithoutRetry timeout" errors.  I don't have a lot of knowledge of the test suites yet...but these are the kinds of places where we may be introducing new flakes.

As for [the e2e-aws-op run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws-op/466):

`level=fatal msg="failed to initialize the cluster: Cluster operator operator-lifecycle-manager has not yet reported success"`

By the time we timed out, looks like the masters had all updated but the worker pool was still in progress:

```
                "machineCount": 3,
                "observedGeneration": 1,
                "readyMachineCount": 2,
                "unavailableMachineCount": 1,
                "updatedMachineCount": 3
```

Looks to me though like all the pivots completed, was probably just finishing by the timeout.

I'm not finding many logs offhand about the operator-lifecycle-manager.

[1] Actually downgraded because we aren't updating `machine-os-content` automatically yet
```
pivot.service: Downgraded:
pivot.service: atomic-openshift-clients 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
pivot.service: atomic-openshift-hyperkube 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
pivot.service: atomic-openshift-node 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
pivot.service: cri-o 1.12.5-6.rhaos4.0.git80d1487.el7 -> 1.12.5-5.rhaos4.0.git9076a33.el7
```

cgwalters commented:
/retest

ashcrow commented:
/test e2e-aws

cgwalters commented:
> [ ](/openshift-ci-robot) ** ci/prow/e2e-aws-op ** — Job succeeded.

:heavy_check_mark: 

But then yeah, [that e2e-aws run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1721) has a pile of failures...some similar to the previous, some new.

ashcrow commented:
There does seem to be a lot of flakes in  `e2e-aws` right now in general.

cgwalters commented:
> There does seem to be a lot of flakes in `e2e-aws` right now in general.

This morning a lot of things fell over due to the bucket limit; there are the random *provisioning* errors at install time but I haven't seen quite as many things failing the actual origin e2e.   I think anyways...

AIUI one thing we can do is [look at the test grid](https://testgrid.k8s.io/redhat-openshift-release-blocking#redhat-release-openshift-origin-installer-e2e-aws-4.0&exclude-non-failed-tests=) and see which of those failed tests are known flaky.

cgwalters commented:
Looking at some other PRs you are right, definitely seeing some of the same issues there. 

I pointed Clayton at this PR...he commented that he thought we were planning to do "early pivot" before the nodes even joined the cluster.  That code is in https://github.com/openshift/machine-config-operator/pull/324

And actually a lot of the infrastructure neccessary for it landed - I'll take a look at rebasing.

cgwalters commented:
@jlebon does this look like all we'd need now?

```
diff --git a/pkg/server/bootstrap_server.go b/pkg/server/bootstrap_server.go
index a163fdd..e8f3091 100644
--- a/pkg/server/bootstrap_server.go
+++ b/pkg/server/bootstrap_server.go
@@ -96,7 +96,7 @@ func (bsc *bootstrapServer) GetConfig(cr poolRequest) (*ignv2_2types.Config, err
 		return nil, fmt.Errorf("server: could not unmarshal file %s, err: %v", fileName, err)
 	}
 
-	appenders := getAppenders(cr, currConf, bsc.kubeconfigFunc)
+	appenders := getAppenders(cr, currConf, bsc.kubeconfigFunc, mc.Spec.OSImageURL)
 	for _, a := range appenders {
 		if err := a(&mc.Spec.Config); err != nil {
 			return nil, err
diff --git a/pkg/server/cluster_server.go b/pkg/server/cluster_server.go
index 82a19b0..76c9692 100644
--- a/pkg/server/cluster_server.go
+++ b/pkg/server/cluster_server.go
@@ -70,7 +70,7 @@ func (cs *clusterServer) GetConfig(cr poolRequest) (*ignv2_2types.Config, error)
 		return nil, fmt.Errorf("could not fetch config %s, err: %v", currConf, err)
 	}
 
-	appenders := getAppenders(cr, currConf, cs.kubeconfigFunc)
+	appenders := getAppenders(cr, currConf, cs.kubeconfigFunc, mc.Spec.OSImageURL)
 	for _, a := range appenders {
 		if err := a(&mc.Spec.Config); err != nil {
 			return nil, err
diff --git a/pkg/server/server.go b/pkg/server/server.go
index dcf8293..9ca64f5 100644
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -16,6 +16,10 @@ const (
 	// of the KubeConfig file on the machine.
 	defaultMachineKubeConfPath = "/etc/kubernetes/kubeconfig"
 
+	// From https://github.com/openshift/pivot/pull/25/commits/c77788a35d7ee4058d1410e89e6c7937bca89f6c#diff-04c6e90faac2675aa89e2176d2eec7d8R44
+	pivotTargetPath = "/etc/pivot/image-pullspec"
+	pivotRebootNeeded = "/run/pivot-reboot-needed"
+
 	// defaultFileSystem defines the default file system to be
 	// used for writing the ignition files created by the
 	// server.
@@ -34,16 +38,26 @@ type Server interface {
 	GetConfig(poolRequest) (*ignv2_2types.Config, error)
 }
 
-func getAppenders(cr poolRequest, currMachineConfig string, f kubeconfigFunc) []appenderFunc {
+func getAppenders(cr poolRequest, currMachineConfig string, f kubeconfigFunc, osimageurl string) []appenderFunc {
 	appenders := []appenderFunc{
 		// append machine annotations file.
 		func(config *ignv2_2types.Config) error { return appendNodeAnnotations(config, currMachineConfig) },
+		// append pivot
+		func(config *ignv2_2types.Config) error { return appendInitialPivot(config, osimageurl) },
 		// append kubeconfig.
 		func(config *ignv2_2types.Config) error { return appendKubeConfig(config, f) },
 	}
 	return appenders
 }
 
+func appendInitialPivot(conf *ignv2_2types.Config, osimageurl string) error {
+	if osimageurl != "" {
+		appendFileToIgnition(conf, pivotTargetPath, osimageurl + "\n")
+		appendFileToIgnition(conf, pivotRebootNeeded, "yes")
+	}
+	return nil
+}
+
 func appendKubeConfig(conf *ignv2_2types.Config, f kubeconfigFunc) error {
 	kcData, _, err := f()
 	if err != nil {
```

jlebon commented:
> pivotRebootNeeded = "/run/pivot-reboot-needed"

`/run/pivot/image-pullspec` (https://github.com/openshift/pivot/commit/d3ba76de7338dc50b7b0ab8856ac2d204baf9ad9)

So one gotcha here is that Ignition will actually write `/run/pivot/image-pullspec` to the filesystem rather than the `/run` tmpfs. So when testing this, I ended up doing:

```json
  "systemd": {
    "units": [
      {
        "contents": "[Unit]\nDescription=Touch /run/pivot/reboot-needed\nConditionFirstBoot=true\n\n[Service]\nType=oneshot\nExecStart=/usr/bin/mkdir /run/pivot\nExecStart=/usr/bin/touch /run/pivot/reboot-needed\n\n[Install]\nWantedBy=multi-user.target\n",
        "enabled": true,
        "name": "pivot-write-reboot-needed.service"
      }
    ]
  }
```

We can clean that up though and just put it in `/etc`. I hadn't prioritized it before since I didn't think we'd end up using this path. (Though actually, having it in `/run` means that we don't have to have special logic in the MCD to not always try to write it out so it can still be in control of reboots).

cgwalters commented:
/test e2e-aws

cgwalters commented:
Hmm...Ignition doesn't seem to have support for writing out systemd units into `/run` does it?

jlebon commented:
> Hmm...Ignition doesn't seem to have support for writing out systemd units into `/run` does it?

Heh, it actually knows how to do this (https://github.com/coreos/ignition/pull/569/commits/af98f37ffdebff367094d3fdf3fe0543cb6b6165), but it's not exposed in the spec, so there's no way to make use of it.

So the unit file itself is in `/etc`.

cgwalters commented:
OK I took a stab at early pivot in https://github.com/openshift/machine-config-operator/pull/426 as a separate PR so we can evaluate them separately.

cgwalters commented:
/retest

cgwalters commented:
One thing we could consider doing with this approach is having the MCO report `available=False` until all nodes have updated at least during initial install.  That would delay e2e runs.

cgwalters commented:
/test e2e-aws-op

runcom commented:
> One thing we could consider doing with this approach is having the MCO report `available=False` until all nodes have updated at least during initial install. That would delay e2e runs.

sounds right to me

ashcrow commented:
/test e2e-aws

cgwalters commented:
I think at this point we're blocking on the e2e-aws tests known to work reliably.  Only then can we gauge our impact on top of that.

jlebon commented:
So testing this on an AWS cluster, I'm getting:

```
I0214 16:51:50.383263   18622 daemon.go:962] While getting MachineConfig master-4548e878b065cb35d3e9690978c90c1b, got: machineconfigs.machineconfiguration.openshift.io "master-4548e878b065cb35d3e9690978c90c1b" not found. Retrying...
```

which sounds like https://github.com/openshift/machine-config-operator/issues/301#issuecomment-455804265. The controller shows:

```
...
I0214 16:49:37.791853       1 template_controller.go:111] Starting MachineConfigController-TemplateController
I0214 16:49:37.891975       1 kubelet_config_controller.go:138] Starting MachineConfigController-KubeletConfigController
I0214 16:49:38.291634       1 render_controller.go:112] Starting MachineConfigController-RenderController
I0214 16:49:38.391709       1 node_controller.go:121] Starting MachineConfigController-NodeController
I0214 16:49:44.791741       1 node_controller.go:383] Pool master is unconfigured, pausing 5s for renderer to initialize
I0214 16:49:44.890824       1 node_controller.go:383] Pool worker is unconfigured, pausing 5s for renderer to initialize
I0214 16:49:46.090966       1 render_controller.go:456] Generated machineconfig worker-0e6bd9375971544777f792fd6e57775d from 3 configs: [{MachineConfig  00-worker  machineconfiguration.openshift.io/v1  } {MachineConfig  00-worker-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-worker-kubelet  machineconfiguration.openshift.io/v1  }]
I0214 16:49:46.190921       1 render_controller.go:456] Generated machineconfig master-bddf086755cc3713f443a66d56fbd5d0 from 3 configs: [{MachineConfig  00-master  machineconfiguration.openshift.io/v1  } {MachineConfig  00-master-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-master-kubelet  machineconfiguration.openshift.io/v1  }]
...
```

Shouldn't we be injecting the osImageURL in the bootstrap path as well here so that checksums match?

jlebon commented:
Forgot to add another blurb after the controller logs:

So the controller isn't regenerating multiple times. Just once per pool, with all the base configs already in place. So I don't think this is #338, right?

jlebon commented:
OK, so it *did* actually work for workers. :fireworks: 

```
I0214 16:52:49.719422    5518 daemon.go:466] In bootstrap mode
I0214 16:52:49.733498    5518 daemon.go:494] Current+desired config: worker-0e6bd9375971544777f792fd6e57775d
I0214 16:52:49.736818    5518 daemon.go:584] Bootstrap pivot required
I0214 16:52:49.736885    5518 update.go:650] Updating OS to registry.svc.ci.openshift.org/rhcos/maipo@sha256:61dc8...
...
rpm-ostreed.service: Txn UpdateDeployment on /org/projectatomic/rpmostree1/redhat_coreos_maipo successful
pivot.service: Upgraded:
pivot.service: nss-altfiles 0-2.atomic.git20131217gite2a80593.el7 -> 2.18.1-11.el7
pivot.service: Downgraded:
pivot.service: bind-libs 32:9.9.4-73.el7_6 -> 32:9.9.4-72.el7
pivot.service: bind-libs-lite 32:9.9.4-73.el7_6 -> 32:9.9.4-72.el7
pivot.service: bind-license 32:9.9.4-73.el7_6 -> 32:9.9.4-72.el7
pivot.service: bind-utils 32:9.9.4-73.el7_6 -> 32:9.9.4-72.el7
...
I0214 16:54:09.828849    5518 update.go:662] machine-config-daemon initiating reboot: Node will reboot into config worker-0e6bd9375971544777f792fd6e57775d
```

Seems like only masters failed with the above.

cgwalters commented:
> Shouldn't we be injecting the osImageURL in the bootstrap path as well here so that checksums match?

We do, that's what https://github.com/openshift/installer/pull/1149 did right?

It did work here above; do you have both commits and an updated installer?

cgwalters commented:
/test e2e-aws-op

kikisdeliveryservice commented:
I'm going to give this a whirl on aws. Will report back.

jlebon commented:
> We do, that's what [openshift/installer#1149](https://github.com/openshift/installer/pull/1149) did right?

OK, right.

> It did work here above; do you have both commits and an updated installer?

Yeah, this is with git master and this PR. Let me dig a little deeper.

cgwalters commented:
In [that last e2e-aws job](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1769) we did an update OK:

```
                "extension": {
                    "master": "all 3 nodes are at latest configuration master-b0dfd7bfa28d55f13cd25bfa1a2bb1a9",
                    "worker": "all 3 nodes are at latest configuration worker-ba41d1dc9e1f87fb672583ba63a1cd1c"
                },
```

kikisdeliveryservice commented:
I applied the controller image and am seeing in controller logs:
`I0214 19:43:01.633722       1 template_controller.go:298] Error syncing controllerconfig machine-config-controller: failed to create MachineConfig for role master: failed to execute template: template: /etc/mcc/templates/master/00-master/_base/files/etc-kubernetes-manifests-etcd-member.yaml:56:29: executing "/etc/mcc/templates/master/00-master/_base/files/etc-kubernetes-manifests-etcd-member.yaml" at <etcdPeerCertDNSNames...>: error calling etcdPeerCertDNSNames: invalid configuration`


this is with installer master

cgwalters commented:
/retest

jlebon commented:
OK, retesting this now with *both* commits in the PR.

runcom commented:
```
NAME                                      GENERATEDBYCONTROLLER   IGNITIONVERSION   CREATED   OSIMAGEURL
00-master                                 3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
00-master-ssh                             3.11.0-616-g60d70a6e                      24m       
00-worker                                 3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
00-worker-ssh                             3.11.0-616-g60d70a6e                      24m       
01-master-kubelet                         3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
01-worker-kubelet                         3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
master-11cf1ace911fdb6293dd9c4d4053f2e6   3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
worker-6f6bf39dcd969f2cb63ea95fb2fff40f   3.11.0-616-g60d70a6e    2.2.0             24m       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:deaee0eeea9cbc2a38381964f0295e0b1e0dda92e4ee3cd46c2b1822da938b0a
```

runcom commented:
upgrades using https://github.com/openshift/machine-config-operator/issues/183#issuecomment-463668093 work like a charm for me

runcom commented:
my cluster looks stable as well, can run apps (ruby), clusteroperators are reporting fine as well. Let's see if e2e-aws reports back...

cgwalters commented:
/hold cancel

I think we should land this.  It feels like the e2e-aws failures aren't related.

jlebon commented:
OK, let's do this! We've seen enough robot runs to gain confidence this works.

/lgtm

cgwalters commented:
After this lands though we will need to quickly set up automated pushes of `machine-os-content`.  Though a plus side of *not* having that right now is that the PR tests here *always* downgrade so we are testing it.

cgwalters commented:
/retest

runcom commented:
tested again, to be also sure my cluster was downgrading at boot (forgot to check in my previous run). Basically bundled the `machine-os-content` from this PR and boom, cluster downgrades:
```
pivot.service: Downgraded:
pivot.service: atomic-openshift-clients 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
pivot.service: atomic-openshift-hyperkube 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
pivot.service: atomic-openshift-node 4.0.0-0.171.0.git.0.a532756.el7 -> 4.0.0-0.170.0.git.0.45dbe92.el7
rpm-ostreed.service: client(id:cli dbus:1.74 unit:pivot.service uid:0) vanished; remaining=0
pivot.service: cri-o 1.12.5-6.rhaos4.0.git80d1487.el7 -> 1.12.5-5.rhaos4.0.git9076a33.el7
pivot.service: Run "systemctl reboot" to start a reboot
```
went from 47.315 to the oscontainer in this PR in `machine-os-content`

runcom commented:
/retest

runcom commented:
/retest

runcom commented:
/retest

runcom commented:
/retest

runcom commented:
/retest

runcom commented:
/retest

ashcrow commented:
/test e2e-aws

runcom commented:
/retest

cgwalters commented:
FWIW...and I could be wrong, I don't think the PR context matters here anymore.  I think [Tide runs its own tests](https://deck-ci.svc.ci.openshift.org/pr?query=is%3Apr+repo%3Aopenshift%2Fmachine-config-operator+author%3Acgwalters+head%3Aosimage-render-but-disabled) and it automatically retests.

jlebon commented:
FWIW, I was finally able to test this properly myself as well. All nodes pivoted on first boot as expected!
Nice work! :confetti_ball:

cgwalters commented:
One thing I'd thought was true but I wanted to dig through the logs to verify is:

I believe we're completing updates during the install, before the e2e tests start.  From [that last e2e-aws run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1849) see e.g. from `machineconfigpools.json`:

```
                        "lastTransitionTime": "2019-02-15T16:19:27Z",
                        "reason": "All nodes are updated with worker-6c6c3c661d5ab71c65bc64d4f03fc066",
                        "lastTransitionTime": "2019-02-15T16:08:25Z",
                        "reason": "All nodes are updated with master-2def176c128ac52721d107243aaed8a5",
```

And:

```
2019/02/15 15:47:34 Running pod e2e-aws
2019/02/15 16:26:13 Container setup in pod e2e-aws completed successfully
```

runcom commented:
> I believe we're completing updates during the install

yes, I can see the lag at cluster bringup where the installer CVO waits for the MCO clusteroperator to be available and as soon as the installation finishes, oc get pods shows the daemons restarted from the update

cgwalters commented:
> FWIW...and I could be wrong, I don't think the PR context matters here anymore. I think [Tide runs its own tests](https://deck-ci.svc.ci.openshift.org/pr?query=is%3Apr+repo%3Aopenshift%2Fmachine-config-operator+author%3Acgwalters+head%3Aosimage-render-but-disabled) and it automatically retests.

Wait, I think I was wrong: 
> ** tide ** Pending — Not mergeable. Job ci/prow/e2e-aws has not succeeded.

/retest

runcom commented:
/retest

cgwalters commented:
Hmm, another consistent failure of those two tests.

cgwalters commented:
OK looking at [the test grid](https://testgrid.k8s.io/redhat-openshift-release-blocking#redhat-release-openshift-origin-installer-e2e-aws-4.0&exclude-non-failed-tests=&sort-by-flakiness=10), `operator Run template e2e-aws - e2e-aws container test 30m22s` is way up there in flakes.

However, the `openshift-tests [Feature:Prometheus][Conformance] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Suite:openshift/conformance/parallel/minimal] 2m40s` doesn't seem to be very flaky.

runcom commented:
OMG the CI doesn't really want to merge this patch

/retest

cgwalters commented:
> OMG the CI doesn't really want to merge this patch

Other PRs are going in though.  I am still worried that something we're doing in the updates is affecting the cluster or later tests.

I'm not seeing a consistent pattern yet though.

ashcrow commented:
HAProxy and Prometheus flakes

/retest

cgwalters commented:
Hmm.  Other PRs are merging...still worried there's something "residual" we're doing to the cluster.  But I just logged into this current e2e-aws cluster, and it looks fine...`oc get clusteroperator` is all clean, same for `oc get pods --all-namespaces`.

cgwalters commented:
Well that last run was a huge set of failures.  Reading the logs, the pools and the operator are reporting ready/done before everything is updated, because they key off currentConfig which we set quickly.  I think we should change the MCD to only set currentConfig when it's done pivoting.  Otherwise the pools are lying.

But even then, the systems seem to be updated.  E.g one of the master MCDs says:
`I0215 22:24:35.343092    6783 daemon.go:647] In desired config master-14bca73029fbcab80e7b5d11d7f131b9`

And the tests start much later:

`2019/02/15 22:35:25 Container setup in pod e2e-aws completed successfully`

cgwalters commented:
/retest

kikisdeliveryservice commented:
@cgwalters there are some bigger CI fixes that are in progress (in other repos for e2e-aws) that should be going in in the next few days.  We can always hold this until you are back on Tuesday and see if the CI resolves itself.

ashcrow commented:
/retest

cgwalters commented:
Looks like [that last run](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1899) hit [this](https://github.com/openshift/machine-config-operator/pull/442).

cgwalters commented:
```
I0216 05:45:03.659689    6232 start.go:52] Version: 3.11.0-670-gb2bebde8-dirty
I0216 05:45:03.660529    6232 start.go:88] starting node writer
I0216 05:45:03.664584    6232 run.go:22] Running captured: chroot /rootfs rpm-ostree status --json
I0216 05:45:03.749243    6232 daemon.go:168] Booted osImageURL: registry.svc.ci.openshift.org/rhcos/maipo@sha256:660061d6eae3ee6d93ca836cd52e6033f1d611c629c1ce47cf272c9e9bda2488 (47.318)                        
I0216 05:45:03.749539    6232 daemon.go:240] Managing node: ip-10-0-156-180.ec2.internal
F0216 05:45:03.770509    6232 start.go:142] binding pod mounts: exec: "mount": executable file not found in $PATH
```

wha :thinking:

cgwalters commented:
/retest

runcom commented:
network failures in the last run (Haproxy being always there)

cgwalters commented:
/test images
/test e2e-aws

runcom commented:
I wonder how https://github.com/openshift/machine-config-operator/pull/442 impacts this PR as well, tests seem to be quite stable there now (especially e2e-aws despite the usual flakes).

smarterclayton commented:
On the mount thing you somehow caught the new 4.0 base image (based on UBI and has no util-linux).

https://github.com/openshift/machine-config-operator/pull/445

will fix your issue and also allow that to get pulled.

smarterclayton commented:
I think this is working because 
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1957 failed with the cri-o bug that they supposedly fixed?

runcom commented:
> I think this is working because
> https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/363/pull-ci-openshift-machine-config-operator-master-e2e-aws/1957 failed with the cri-o bug that they supposedly fixed?

yeah, that should be fixed assuming we're testing with the right crio package /cc @mrunalp

smarterclayton commented:
Yeah, machine-os-content was 313 which didn't have the fix.  I just bumped to 318

/retest

cgwalters commented:
OK my view of status here.  First...I think I fixed https://github.com/openshift/machine-config-operator/pull/426 and I'd like to go with that (which is an additional commit on top) because this version isn't correctly having the controller manage updates, basically a node comes up and joins the cluster, MCD lands then we immediately reboot - see that happening rapidly for the workers e.g.

> I think this is working

It's *definitely* downgrading the cluster.  If some of these e2e failures come down to e.g. tests assuming slightly newer kubelet or cri-o then indeed we should bump `machine-os-content` again.

cgwalters commented:
/lgtm cancel

Since I'd like to keep trying for https://github.com/openshift/machine-config-operator/pull/426
(but it's OK if this goes in as the other is a superset)

cgwalters commented:
/hold

cgwalters commented:
Closing this one in favor of https://github.com/openshift/machine-config-operator/pull/426 which includes it.
